### ISSUE TITLE
InvalidRequirement when installing wheel with URL in requirements

### ISSUE THREAD

**Issue by Timmmm** [pypa/pipenv#4640]
### Issue description

I have a wheel file with the following in its metadata:

```
Requires-Dist: torch @ https://download.pytorch.org/whl/cpu/torch-1.7.1%2Bcpu-cp36-cp36m-linux_x86_64.whl
```

This is a valid requirements string according to [the grammar](https://www.python.org/dev/peps/pep-0508/#grammar) (see `urlspec`).

### Expected result

When I run `pipenv install file:///path/to/thewheel.whl` it should work.

### Actual result

It gives this error:

```
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/vendor/packaging/requirements.py", line 98, in __init__
    req = REQUIREMENT.parseString(requirement_string)
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/vendor/pyparsing.py", line 1955, in parseString
    raise exc
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/vendor/pyparsing.py", line 3814, in parseImpl
    raise ParseException(instring, loc, self.errmsg, self)
pyparsing.ParseException: Expected stringEnd, found ':'  (at char 5), (line:1, col:6)
```

I hacked up the code to verity that it is failing on the above string (it would be nice if it printed the entire string it was failing on!).

### Steps to replicate

Unfortunately the wheel is private but you can probably just hack an existing one and add `@ https://foo.com/bar`.

-------------------------------------------------------------------------------
<details><summary>$ pipenv --support</summary>

Pipenv version: `'2020.11.15'`

Pipenv location: `'/home/me/.local/lib/python3.6/site-packages/pipenv'`

Python location: `'/usr/bin/python3'`

Python installations found:

  - `3.6.9`: `/usr/bin/python3`
  - `3.6.9`: `/usr/bin/python3.6`
  - `3.6.9`: `/usr/bin/python3.6m`
  - `2.7.15`: `/usr/bin/python2`
  - `2.7.15`: `/usr/bin/python2.7`

PEP 508 Information:

```
{'implementation_name': 'cpython',
 'implementation_version': '3.6.9',
 'os_name': 'posix',
 'platform_machine': 'x86_64',
 'platform_python_implementation': 'CPython',
 'platform_release': '4.15.0-55-generic',
 'platform_system': 'Linux',
 'platform_version': '#60-Ubuntu SMP Tue Jul 2 18:22:20 UTC 2019',
 'python_full_version': '3.6.9',
 'python_version': '3.6',
 'sys_platform': 'linux'}
```

System environment variables:

  - `NVM_DIR`
  - `LS_COLORS`
  - `XDG_CONFIG_HOME`
  - `SSH_CONNECTION`
  - `LESSCLOSE`
  - `LANG`
  - `EDITOR`
  - `RUSTUP_HOME`
  - `COLORTERM`
  - `NVM_CD_FLAGS`
  - `NPM_CONFIG_USERCONFIG`
  - `S_COLORS`
  - `XDG_SESSION_ID`
  - `USER`
  - `PWD`
  - `HOME`
  - `BROWSER`
  - `VSCODE_GIT_ASKPASS_NODE`
  - `TERM_PROGRAM`
  - `SSH_CLIENT`
  - `TERM_PROGRAM_VERSION`
  - `VSCODE_IPC_HOOK_CLI`
  - `CLICOLOR_FORCE`
  - `CARGO_HOME`
  - `MAIL`
  - `VSCODE_GIT_ASKPASS_MAIN`
  - `TERM`
  - `SHELL`
  - `NVM_BIN`
  - `SHLVL`
  - `LANGUAGE`
  - `NPM_CONFIG_CACHE`
  - `VSCODE_GIT_IPC_HANDLE`
  - `NOTHOME`
  - `XDG_CACHE_HOME`
  - `LOGNAME`
  - `DBUS_SESSION_BUS_ADDRESS`
  - `GIT_ASKPASS`
  - `XDG_RUNTIME_DIR`
  - `NPM_CONFIG_TMP`
  - `XDG_DATA_HOME`
  - `PATH`
  - `GCC_COLORS`
  - `LESSOPEN`
  - `_`
  - `OLDPWD`
  - `PIP_DISABLE_PIP_VERSION_CHECK`
  - `PYTHONDONTWRITEBYTECODE`
  - `PIP_SHIMS_BASE_MODULE`
  - `PIP_PYTHON_PATH`
  - `PYTHONFINDER_IGNORE_UNSUPPORTED`

Pipenv–specific environment variables:

Debug–specific environment variables:

  - `PATH`: `redacted:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games`
  - `SHELL`: `/bin/bash`
  - `EDITOR`: `code --wait`
  - `LANG`: `en_GB.utf8`
  - `PWD`: `/home/me/d4`

---------------------------

</details>

**Comment by Timmmm** [comment]
Ok the weird thing is that `requirement_string` is `https://download.pytorch.org/whl/cpu/torch-1.7.1%2Bcpu-cp36-cp36m-linux_x86_64.whl`, not `torch @ https://download.pytorch.org/whl/cpu/torch-1.7.1%2Bcpu-cp36-cp36m-linux_x86_64.whl` which looks like it would parse fine.

Also I think the very first time I did this it did work. Possibly a caching issue. Not sure if Pipenv has a cache anywhere.

**Comment by Timmmm** [comment]
Here's the full stack trace (Python's stack trace output is kind of confusing):

```
Traceback (most recent call last):
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/resolver.py", line 764, in <module>
    main()
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/resolver.py", line 760, in main
    dev=parsed.dev)
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/resolver.py", line 741, in _main
    resolve_packages(pre, clear, verbose, system, write, requirements_dir, packages, dev)
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/resolver.py", line 709, in resolve_packages
    requirements_dir=requirements_dir,
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/resolver.py", line 692, in resolve
    req_dir=requirements_dir
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/utils.py", line 1403, in resolve_deps
    req_dir=req_dir,
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/utils.py", line 1108, in actually_resolve_deps
    resolver.resolve()
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/utils.py", line 823, in resolve
    results = self.resolver.resolve(max_rounds=environments.PIPENV_MAX_ROUNDS)
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/patched/piptools/resolver.py", line 210, in resolve
    reverse_dependencies = self.reverse_dependencies(results)
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/patched/piptools/resolver.py", line 404, in reverse_dependencies
    return self.dependency_cache.reverse_dependencies(non_editable)
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/patched/piptools/cache.py", line 143, in reverse_dependencies
    return self._reverse_dependencies(ireqs_as_cache_values)
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/patched/piptools/cache.py", line 168, in _reverse_dependencies
    for name, version_and_extras in cache_keys
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/patched/piptools/utils.py", line 325, in lookup_table
    for value in values:
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/patched/piptools/cache.py", line 169, in <genexpr>
    for dep_name in self.cache[name][version_and_extras]
  File "/home/me/.local/lib/python3.6/site-packages/pipenv/vendor/packaging/requirements.py", line 102, in __init__
    requirement_string, e.msg, requirement_string
```

**Comment by Timmmm** [comment]
Ok I figured it out. In `resolver.py` it runs this code:

            self.dependency_cache[ireq] = sorted(set(format_requirement(ireq) for ireq in dependencies))

In `format_requirement()` there's this line:

    elif is_url_requirement(ireq):
        line = ireq.link.url

`is_url_requirement` returns true, so `line` is set to just `https://download.pytorch.org/whl/cpu/torch-1.7.1%2Bcpu-cp36-cp36m-linux_x86_64.whl` rather than `torch @ https://download.pytorch.org/whl/cpu/torch-1.7.1%2Bcpu-cp36-cp36m-linux_x86_64.whl`.

That gets added to the cache, and then later we come to this code

        return lookup_table(
            (key_from_req(Requirement(dep_name)), name)
            for name, version_and_extras in cache_keys
            for dep_name in self.cache[name][version_and_extras]
        )

which calles `Requirement("https://download.pytorch.org/whl/cpu/torch-1.7.1%2Bcpu-cp36-cp36m-linux_x86_64.whl")` resulting in the parse error.

Pretty difficult to say how this is supposed to work because it's all stringly typed and therefore really hard to follow. Although I did see that `_requirement_to_str_lowercase_name` already handles the URL? So maybe just deleting these lines should work?

    elif is_url_requirement(ireq):
        line = ireq.link.url

I tried it but then I get this error:

```
TypeError: Expected pinned or editable requirement, got torch@ https://download.pytorch.org/whl/cpu/torch-1.7.1%2Bcpu-cp36-cp36m-linux_x86_64.whl
```

**Comment by matteius** [comment]
@Timmmm Can you please re-test with `2022.1.8`?

**Comment by matteius** [comment]
Sorry, the version to test with now is `2022.4.8`
